src/Http/Routing/src/Matching/HeaderMatcherPolicy.cs

```diff
@@ -0,0 +1,207 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal sealed class HeaderMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, IEndpointSelectorPolicy
+    {
+        /// <inheritdoc/>
+        // Run after HttpMethodMatcherPolicy (-1000) and HostMatcherPolicy (-100), but before default (0)
+        public override int Order => -50;
+
+        /// <inheritdoc/>
+        public IComparer<Endpoint> Comparer => new HeaderMetadataEndpointComparer();
+
+        /// <inheritdoc/>
+        bool IEndpointSelectorPolicy.AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            _ = endpoints ?? throw new ArgumentNullException(nameof(endpoints));
+
+            // When the node contains dynamic endpoints we can't make any assumptions.
+            if (ContainsDynamicEndpoints(endpoints))
+            {
+                return true;
+            }
+
+            return AppliesToEndpointsCore(endpoints);
+        }
+
+        /// <inheritdoc/>
+        public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
+        {
+            _ = httpContext ?? throw new ArgumentNullException(nameof(httpContext));
+            _ = candidates ?? throw new ArgumentNullException(nameof(candidates));
+
+            for (int i = 0; i < candidates.Count; i++)
+            {
+                if (!candidates.IsValidCandidate(i))
+                {
+                    continue;
+                }
+
+                var metadata = candidates[i].Endpoint.Metadata.GetMetadata<IHeaderMetadata>();
+                var metadataHeaderName = metadata?.HeaderName;
+                if (string.IsNullOrEmpty(metadataHeaderName))
+                {
+                    // Can match any request
+                    continue;
+                }
+
+                var metadataHeaderValues = metadata.HeaderValues;
+                int maxValuesToInspect = metadata.MaximumValuesToInspect;
+
+                var matched = false;
+                if (httpContext.Request.Headers.TryGetValue(metadataHeaderName, out var requestHeaderValues))
+                {
+                    if (metadataHeaderValues?.Count == 0)
+                    {
+                        // We were asked to match as long as the header exists, and it *does* exist
+                        matched = true;
+                    }
+                    else
+                    {
+                        var comparisonFunc = GetComparisonFunc(metadata.ValueMatchMode);
+                        var stringComparison = metadata.ValueIgnoresCase ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
+                        for (int j = 0; j < metadataHeaderValues.Count; j++)
+                        {
+                            for (int k = 0; k < requestHeaderValues.Count && k < maxValuesToInspect; k++)
```

Comment:
Reposing my format feedback, I think it was obscured by another change.

There's a header formatting issue here.

Officially HTTP considers the following examples to be equivalent:
```
Header: value1
Header: value2
Header: value3
```
```
Header: value1, value2, value3
```
```
Header: value1, value2
Header: value3
```

What Headers.TryGetValue returns is a raw list of the headers as received from the client, it doesn't account for splitting or merging equivalency. You could get any of the above formats. You're matching the config value vs a single header array entry, but that entry could still contain multiple values.

There's an extension `GetCommaSeparatedValues` that splits up the headers into a unified format (split by comma, remove quotes).
https://github.com/dotnet/aspnetcore/blob/aedcfbd5c0350ad7dd1e882cafbaf692b979b2e8/src/Http/Http.Abstractions/src/Extensions/HeaderDictionaryExtensions.cs#L39-L42
We can have a separate discussion about its efficiency, but for now it will address your correctness issue.